### PR TITLE
Leverage Enum properties provided by AWS

### LINKED ISSUE #236: Handle Enum
The api2-json file provides hints on enum. (ie. the Runtime Type in in Lambda)

We should:
- provide const for the values
- take those value in "validators"

```php
class ListLayerVersionsRequest
{
    /**
     * A runtime identifier. For example, `go1.x`.
     *
     * @var string|null
     */
    private $CompatibleRuntime;

    public function validate(): void
    {
        foreach ($this->CompatibleRuntime as $value) {
            if (!isset(Runtime::AVAILABLE_RUNTIMES[$value])) {
                throw new InvalidArgument(sprintf('The CompatibleRuntime "%s" in "%s" is not a Valide "Runtime". Available values are "%"', $value, __CLASS__, array_keys(Runtime::AVAILABLE_RUNTIMES)));
            }
        }
    }
}

// Class is named "Runtime" because of the name of the shape
class Runtime 
{
    public const NODEJS = 'nodejs';
    public const NODEJS_4_3 = 'nodejs4.3';
    public const NODEJS_6_10 = 'nodejs6.10';
    public const NODEJS_4_3_EDGE = 'nodejs4.3-edge';
    // ..

    // = "AVAILABLE_" + className.upper()
    public const AVAILABLE_RUNTIME = [
        self::NODEJS = true,
        self::NODEJS_4_3 = true,
        self::NODEJS_6_10 = true,
        self::NODEJS_4_3_EDGE = true,
        // ..
    ];
}
```

### PR DESCRIPTION
**Pull request by jderusse**
fixes #236

### CODE REVIEW

File: src/Service/CloudFormation/src/Result/StackEvent.php [hunk 1]
```diff
@@ -43,6 +45,8 @@ class StackEvent
 
     /**
      * Current status of the resource.
+     *
+     * @see ResourceStatus
```
Comment (jderusse) [inline src/Service/CloudFormation/src/Result/StackEvent.php]: does somebody know how to define enum in dockblock? something like `@var string{'foo'|'bar'}`.

File: src/Service/Lambda/src/Input/PublishLayerVersionRequest.php [hunk 2]
```diff
@@ -196,13 +197,19 @@ public function setLicenseInfo(?string $value): self
...
-                throw new InvalidArgument(sprintf('Missing parameter "%s" when validating the "%s". The value cannot be null.', $name, __CLASS__));
-            }
+        if (null === $this->LayerName) {
+            throw new InvalidArgument(sprintf('Missing parameter "LayerName" when validating the "%s". The value cannot be null.', __CLASS__));
+        }
+
+        if (null === $this->Content) {
+            throw new InvalidArgument(sprintf('Missing parameter "Content" when validating the "%s". The value cannot be null.', __CLASS__));
         }
-        if ($this->Content) {
```
Comment (jderusse) [inline src/Service/Lambda/src/Input/PublishLayerVersionRequest.php]: I also refactor how requires properties are handled in validation: I replace `foreach (['foo', 'bar'` by `if (foo... if (bar...`.

That allows me to group similar validation on same properties and fix useless check like this one.

File: src/Service/S3/tests/Integration/S3ClientTest.php [hunk 3]
```diff
@@ -300,9 +302,9 @@ public function testPutObjectAcl(): void
                 'Grants' => [new Grant([
                     'Grantee' => new Grantee([
                         'DisplayName' => 'me',
-                        'Type' => 'canonical',
+                        'Type' => Type::CANONICALUSER,
```
Comment (jderusse) [inline src/Service/S3/tests/Integration/S3ClientTest.php]: Bug spotted :)

File: src/Service/S3/src/Enum/Type.php [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace AsyncAws\S3\Enum;
+
+class Type
+{
+    public const AMAZONCUSTOMERBYEMAIL = 'AmazonCustomerByEmail';
+    public const CANONICALUSER = 'CanonicalUser';
```
Comment (jderusse) [inline src/Service/S3/src/Enum/Type.php]: should be `CANONICAL_USER`

File: src/Service/S3/tests/Integration/S3ClientTest.php (hunk 3 above)
Comment (Nyholm) [reply src/Service/S3/tests/Integration/S3ClientTest.php]: Are you sure?

File: src/Core/src/Sts/Input/AssumeRoleWithWebIdentityRequest.php [hunk 5]
```diff
@@ -240,11 +240,18 @@ public function setWebIdentityToken(?string $value): self
 
     public function validate(): void
     {
-        foreach (['RoleArn', 'RoleSessionName', 'WebIdentityToken'] as $name) {
-            if (null === $this->$name) {
```
Comment (Nyholm) [inline src/Core/src/Sts/Input/AssumeRoleWithWebIdentityRequest.php]: Did we remove “name”?

File: src/Service/S3/tests/Integration/S3ClientTest.php (hunk 3 above)
Comment (jderusse) [reply src/Service/S3/tests/Integration/S3ClientTest.php]: The const name is generated. So, yes, we ca choose whatever we want :)
Note: the value of the constant remain the same and is the value expected by AWS.

Maybe it's not obvious for `CANONICALUSER`, I found that `AMAZON_CUSTOMER_BY_EMAIL` is most readable than `AMAZONCUSTOMERBYEMAIL`

I fixed it.

Comment (jderusse) [reply src/Service/S3/tests/Integration/S3ClientTest.php]: yes. The right value is `CanonicalUser`

File: src/Service/S3/src/Enum/Type.php (hunk 4 above)
Comment (jderusse) [reply src/Service/S3/src/Enum/Type.php]: fixe

File: src/Core/src/Sts/Input/AssumeRoleWithWebIdentityRequest.php (hunk 5 above)
Comment (jderusse) [reply src/Core/src/Sts/Input/AssumeRoleWithWebIdentityRequest.php]: `"name"`?.. you mean `$name` which is a variable containing either `RoleArn`, `RoleSessionName` or `WebIdentityToken`. The check is still performed bellow

File: src/Service/Lambda/src/Input/PublishLayerVersionRequest.php (hunk 2 above)
Comment (Nyholm) [reply src/Service/Lambda/src/Input/PublishLayerVersionRequest.php]: Cool. I see this now.

File: src/Service/S3/src/Input/PutObjectRequest.php [hunk 6]
```diff
@@ -799,9 +815,47 @@ public function setWebsiteRedirectLocation(?string $value): self
...
+        }
+
+        if (null !== $this->StorageClass) {
+            if (!isset(StorageClass::AVAILABLE_STORAGECLASS[$this->StorageClass])) {
+                throw new InvalidArgument(sprintf('Invalid parameter "StorageClass" when validating the "%s". The value "%s" is not a valid "StorageClass". Available values are %s.', __CLASS__, $this->StorageClass, implode(', ', array_keys(StorageClass::AVAILABLE_STORAGECLASS))));
+            }
+        }
+
+        if (null !== $this->RequestPayer) {
+            if (!isset(RequestPayer::AVAILABLE_REQUESTPAYER[$this->RequestPayer])) {
```
Comment (Nyholm) [inline src/Service/S3/src/Input/PutObjectRequest.php]: Just making an observation: if aws adds something to this enum value, we need to release a new version before someone can use the new aws feature.

File: src/Service/CloudFormation/src/Result/StackEvent.php (hunk 1 above)
Comment (Nyholm) [reply src/Service/CloudFormation/src/Result/StackEvent.php]: Look at the bottom of this page: https://psalm.dev/docs/annotating_code/typing_in_psalm/

File: src/Service/S3/src/Input/PutObjectRequest.php (hunk 6 above)
Comment (jderusse) [reply src/Service/S3/src/Input/PutObjectRequest.php]: Good catch, but I think this is similar for every change in the Api contract (new property, new method, new(or remove) requirement...

Comment (Nyholm) [reply src/Service/S3/src/Input/PutObjectRequest.php]: See #258

File: src/Core/src/Sts/Input/AssumeRoleRequest.php [hunk 7]
```diff
@@ -175,6 +178,9 @@ public function getSerialNumber(): ?string
         return $this->SerialNumber;
     }
 
+    /**
+     * @return Tag[]
```
Comment (jderusse) [inline src/Core/src/Sts/Input/AssumeRoleRequest.php]: Not really related to that PR, but definining such dockbloc helps users and static tools (phpstan/psalm) following variable assignments.

Comment (Nyholm) [reply src/Core/src/Sts/Input/AssumeRoleRequest.php]: Yes. We should do more of these

File: phpstan-baseline.neon [hunk 8]
```diff
@@ -10,3 +10,9 @@ parameters:
 			count: 1
 			path: src/Integration/Symfony/Bundle/src/DependencyInjection/Configuration.php
 
+		-   "#^PHPDoc tag @param has invalid value#"
```
Comment (jderusse) [inline phpstan-baseline.neon]: phpstan don't like the Enum param docblock.. I had to ignore all those rules :/

**Comment by Nyholm** [comment]
Is this ready to merge?

**Comment by jderusse** [comment]
Yes, it's RFR => If your're OK with new PHPStan ignore rules  (cf https://github.com/async-aws/aws/pull/257#discussion_r386133675)

File: phpstan-baseline.neon (hunk 8 above)
Comment (Nyholm) [reply phpstan-baseline.neon]: Im okey with this. Pslam is happy.

**Pull request merged**